File: examples/telemetry/06_exporter_resilience_modes.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""🛡️ Exporter resilience — retries, timeouts, and failure policies.

Demonstrates:
- ExporterPolicy with fail_open=True vs fail_open=False
- timeout_seconds for deadline enforcement
- get_exporter_policy to inspect active policy
- Health snapshot: retries, export failures, latency, last error
"""

from __future__ import annotations

import time

from provide.telemetry import ExporterPolicy, get_exporter_policy, get_health_snapshot, set_exporter_policy
from provide.telemetry.resilience import get_circuit_state, run_with_resilience


def main() -> None:
    print("🛡️  Exporter Resilience Demo\n")

    # ── 🟢 Fail-open: returns None on failure ────────────
    print("🟢 Fail-open mode (retries=1, backoff=0s)")
    set_exporter_policy("logs", ExporterPolicy(retries=1, backoff_seconds=0.0, fail_open=True))

    attempts_open = {"count": 0}

    def flaky_fail_open() -> str:
        attempts_open["count"] += 1
        raise RuntimeError("simulated exporter failure")

    result = run_with_resilience("logs", flaky_fail_open)
    policy = get_exporter_policy("logs")
    print(f"  📦 Result: {result}")
    print(f"  🔄 Attempts: {attempts_open['count']}")
    print(f"  📋 Policy: retries={policy.retries}, fail_open={policy.fail_open}")

    # ── 🔴 Fail-closed: raises on failure ────────────────
    print("\n🔴 Fail-closed mode (retries=1, backoff=0s)")
    set_exporter_policy("logs", ExporterPolicy(retries=1, backoff_seconds=0.0, fail_open=False))

    attempts_closed = {"count": 0}

    def flaky_fail_closed() -> str:
        attempts_closed["count"] += 1
        raise RuntimeError("simulated hard failure")

    try:
        run_with_resilience("logs", flaky_fail_closed)
    except RuntimeError as exc:
        print(f"  💥 Caught: {exc}")
        print(f"  🔄 Attempts: {attempts_closed['count']}")

    # ── ⏱️ Timeout enforcement ────────────────────────────
    print("\n⏱️  Timeout enforcement (timeout=0.05s)")
    set_exporter_policy(
        "traces",
        ExporterPolicy(retries=0, timeout_seconds=0.05, fail_open=True),
    )

    def slow_export() -> str:
        time.sleep(0.2)
        return "too late"

    timed_out = run_with_resilience("traces", slow_export)
    print(f"  📦 Result: {timed_out}  (None = timed out, fail-open)")

    # ── 🔌 Circuit breaker with exponential backoff ─────────
    print("\n🔌 Circuit breaker (exponential backoff + half-open probing)")
    set_exporter_policy(
        "metrics",
        ExporterPolicy(retries=0, timeout_seconds=0.01, fail_open=True),
    )

    def always_timeout() -> str:
        time.sleep(1)
        return "never"

    for _i in range(4):
        run_with_resilience("metrics", always_timeout)

    state, open_count, cooldown_remaining = get_circuit_state("metrics")
    print(f"  🔌 Circuit state:     {state}")
    print(f"  📈 Open count:        {open_count}")
    print(f"  ⏳ Cooldown remaining: {cooldown_remaining:.1f}s")

    # ── 📊 Health snapshot ────────────────────────────────
    print("\n📊 Health snapshot after all operations:")
    snapshot = get_health_snapshot()
    print(f"  🔄 retries_logs:          {snapshot.retries_logs}")
    print(f"  ❌ export_failures_logs:   {snapshot.export_failures_logs}")
    print(f"  ❌ export_failures_traces:  {snapshot.export_failures_traces}")
    print(f"  ⏱️  latency_ms_traces:     {snapshot.export_latency_ms_traces}")
    print(f"  🔌 circuit_state_metrics:  {snapshot.circuit_state_metrics}")
    print(f"  📈 circuit_open_count:     {snapshot.circuit_open_count_metrics}")
    print(f"  🛑 setup_error:           {snapshot.setup_error}")

    print("\n🏁 Done!")


if __name__ == "__main__":
    main()

File: typescript/examples/telemetry/06_exporter_resilience_modes.ts
// SPDX-FileCopyrightText: Copyright (C) 2026 MindTenet LLC
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of Provide Telemetry.

/**
 * 🛡️ Exporter resilience — retries, timeouts, and failure policies.
 *
 * Demonstrates:
 * - ExporterPolicy with failOpen=true vs failOpen=false
 * - timeoutMs for deadline enforcement
 * - getExporterPolicy to inspect active policy
 * - runWithResilience for resilient async exports
 * - Health snapshot: exportRetries, exportFailures, exportLatencyMs, lastExportError
 *
 * Run: npx tsx examples/telemetry/06_exporter_resilience_modes.ts
 */

import {
  getExporterPolicy,
  getHealthSnapshot,
  runWithResilience,
  setExporterPolicy,
  setupTelemetry,
  shutdownTelemetry,
} from '../../src/index.js';

async function main(): Promise<void> {
  console.log('🛡️  Exporter Resilience Demo\n');

  setupTelemetry({ serviceName: 'provide-telemetry-examples', consoleOutput: false });

  // ── 🟢 Fail-open: returns null on failure ────────────
  console.log('🟢 Fail-open mode (retries=1, backoffMs=0)');
  setExporterPolicy({ retries: 1, backoffMs: 0, failOpen: true, timeoutMs: 0 });

  let attemptsOpen = 0;
  const result = await runWithResilience('logs', async () => {
    attemptsOpen++;
    throw new Error('simulated exporter failure');
  });
  const policy = getExporterPolicy();
  console.log(`  📦 Result: ${result}`);
  console.log(`  🔄 Attempts: ${attemptsOpen}`);
  console.log(`  📋 Policy: retries=${policy.retries}, failOpen=${policy.failOpen}`);

  // ── 🔴 Fail-closed: throws on failure ────────────────
  console.log('\n🔴 Fail-closed mode (retries=1, backoffMs=0)');
  setExporterPolicy({ retries: 1, backoffMs: 0, failOpen: false, timeoutMs: 0 });

  let attemptsClosed = 0;
  try {
    await runWithResilience('logs', async () => {
      attemptsClosed++;
      throw new Error('simulated hard failure');
    });
  } catch (err) {
    console.log(`  💥 Caught: ${(err as Error).message}`);
    console.log(`  🔄 Attempts: ${attemptsClosed}`);
  }

  // ── ⏱️ Timeout enforcement ────────────────────────────
  console.log('\n⏱️  Timeout enforcement (timeoutMs=50)');
  setExporterPolicy({ retries: 0, timeoutMs: 50, failOpen: true });

  const timedOut = await runWithResilience('traces', async () => {
    await new Promise((r) => setTimeout(r, 200));
    return 'too late';
  });
  console.log(`  📦 Result: ${timedOut}  (null = timed out, fail-open)`);

  // ── 📊 Health snapshot ────────────────────────────────
  console.log('\n📊 Health snapshot after all operations:');
  const snapshot = getHealthSnapshot();
  console.log(`  🔄 exportRetries:    ${snapshot.exportRetries}`);
  console.log(`  ❌ exportFailures:   ${snapshot.exportFailures}`);
  console.log(`  💬 lastExportError:  ${snapshot.lastExportError}`);
  console.log(`  ⏱️  exportLatencyMs: ${snapshot.exportLatencyMs}`);

  console.log('\n🏁 Done!');
  await shutdownTelemetry();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
